keps/sig-node/2369-sizing-providers/README.md

```diff
@@ -0,0 +1,319 @@
+	APIVersion string `json:"apiVersion"`
+
+	// Arguments to pass to the command when executing it.
+	// +optional
+	Args []string `json:"args,omitempty"`
+
+	// Env defines additional environment variables to expose to the process. These
+	// are unioned with the host's environment, as well as variables client-go uses
+	// to pass argument to the plugin.
+	// +optional
+	Env []ExecEnvVar `json:"env,omitempty"`
+}
+
+// ExecEnvVar is used for setting environment variables when executing an exec-based
+// sizing plugin.
+type ExecEnvVar struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+```
+
+### Sizing Provider Request API
+
+If an exec plugin is enabled the kubelet will exec the plugin during the startup, passing the `SizingProviderRequest` API via stdin.
+
+The kubelet will encode the request based on the apiVersion provided in SizingProviderConfig. It wil also exec the plugin based on the `args` and `env` fields in `SizingProviderConfig`.
+
+```go
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SizingProviderRequest will be passed to the plugin via stdin. In general,
+// plugins should prefer responding with the same apiVersion they were sent.
+type SizingProviderRequest struct {
+	metav1.TypeMeta
+}
+```
+
+### Sizing Provider Response API
+
+An exec plugin is expected to return an encoded response of the `SizingProviderResponse` API to the kubelet via stdout. 
+
+```go
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SizingProviderResponse holds sizing values that the kubelet should use for the specified
+// image provided in the original request. Kubelet will read the response from the plugin via stdout.
+type SizingProviderResponse struct {
+	metav1.TypeMeta
+ 
+      // A set of ResourceName=ResourceQuantity (e.g. cpu=200m,memory=150G,pid=100) pairs
+	// that describe resources reserved for non-kubernetes components.
+	// Currently only cpu and memory are supported.
+	// See http://kubernetes.io/docs/user-guide/compute-resources for more detail.
+	SystemReserved map[string]string `json:"systemReserve"`
+	// A set of ResourceName=ResourceQuantity (e.g. cpu=200m,memory=150G,pid=100) pairs
+	// that describe resources reserved for kubernetes system components.
+	// Currently cpu, memory and local ephemeral storage for root file system are supported.
+	// See http://kubernetes.io/docs/user-guide/compute-resources for more detail.
+	KubeReserved map[string]string `json:"kubeReserve"`
+}
+
+```
+
+### Test Plan
+
+Alpha:
+* unit tests for the exec plugin provider
+* unit tests for API validation
+
+### Graduation Criteria
+
+### Alpha
+
+* adequate unit testing for the plugin provider
+* a working reference implementation, proving that the existing functionality of the built-in providers
+can be achieved using the exec plugin.
+
+#### Alpha -> Beta Graduation
+
+* integration or e2e tests.
+* at least one working plugin implementation.
```

Comment:
two metrics worth tracking for exec plugins were defined as part of kubelet cred provider plugin.

see:
https://github.com/kubernetes/enhancements/pull/2457/files#diff-7664c031fa927970c5fc47a1a3e1991e6691918066576490797650aaced971d3R316

similar metrics would apply here.

Reply:
There is a slight difference between `cred provider plugin` and this `node sizing provider plugin`. 

The `cred provider plugin` gets executed throughout the lifetime of the kubelet whenever there is a request to provision a pod with matching image name. So It makes sense for them to have `kubelet_credential_provider_plugin_errors` which will track the number errors that occurred from invoking an exec plugin and `kubelet_credential_provider_plugin_duration` which will track the duration of execution by plugins.

On the other hand, `node sizing provider plugin` gets executed only during the start up of the kubelet. Unlike `cred provider plugin`, there is no repeated execution of `node sizing provider plugin` by the kubelet spread over the time (as long as kubelet is running). If a ` node sizing provider plugin` is enabled and it fails to fetch the sizing values, the kubelet will abort the startup and log the error in kubelet logs. Since the user has explicitly requested the sizing values to be fetched using the exec plugin, kubelet should not default to any other values. Defaulting to any other values and bringing up the kubelet may result in an unpredictable behavior, such as node lock ups, as the default sizing values may be not optimal. 

The `node sizing provider plugin` is either going to succeed or fail during the kubelet startup, since it's not going to generate any time series data like `cred provider plugin` I have deliberately dropped the `Metrics` section from this KEP. 

But in case you still feel that there is a scope for adding any other metrics, please let me know.